File: examples/python/03_payments.py
"""Scenarios 8-11: build a kingpepe: URI, monitor an address, wait for
confirmations, and send a payment safely. Use regtest for development."""

from __future__ import annotations

import time

from kingpepe_sdk import (
    KingPepeClient,
    check_payment,
    create_payment_request,
    parse_kpepe,
    payment_uri,
)


def main() -> None:
    client = KingPepeClient.from_env()
    wallet = client.with_wallet("example-shop")

    # 8) Create a payment request and its kingpepe: URI.
    req = create_payment_request(
        id=f"order-{int(time.time())}",
        address=wallet.get_new_address("checkout"),
        amount_base_units=parse_kpepe("1.25"),
        required_confirmations=1,
        ttl_ms=30 * 60 * 1000,
    )
    print("Pay to:", payment_uri(req))

    # 11) (demo) On regtest, pay it ourselves so the example is self-contained.
    #     In production the buyer pays; do NOT send from your own wallet.
    if client.network == "regtest":
        txid = wallet.send_to_address(req.address, req.amount_base_units)
        print("Sent demo payment:", txid)
        wallet.generate_to_address(1, wallet.get_new_address())  # confirm it

    # 9-10) Poll status until terminal (or a few iterations).
    state = None
    for _ in range(10):
        state = check_payment(wallet, req)
        print(f"status={state.status} confirmations={state.observation.confirmations}")
        if state.terminal:
            break
        time.sleep(2)
    print("Final status:", state.status if state else "unknown")


if __name__ == "__main__":
    main()

File: examples/typescript/03_payments.ts
// Scenarios 8-11: build a kingpepe: URI, monitor an address, wait for
// confirmations, and send a payment safely. Use regtest for development.
import {
  KingPepeClient,
  createPaymentRequest,
  paymentUri,
  checkPayment,
  parseKpepe,
  type PaymentState,
} from "@kingpepe/sdk";

async function main() {
  const client = KingPepeClient.fromEnv();
  const wallet = client.withWallet("example-shop");

  // 8) Create a payment request and its kingpepe: URI.
  const req = createPaymentRequest({
    id: `order-${Date.now()}`,
    address: await wallet.getNewAddress("checkout"),
    amountBaseUnits: parseKpepe("1.25"),
    requiredConfirmations: 1,
    ttlMs: 30 * 60 * 1000,
  });
  console.log("Pay to:", paymentUri(req));

  // 11) (demo) pay it ourselves so the example is self-contained on regtest.
  //     In production the buyer pays; you would NOT send from your own wallet.
  if (client.network === "regtest") {
    const txid = await wallet.sendToAddress(req.address, req.amountBaseUnits);
    console.log("Sent demo payment:", txid);
    await wallet.generateToAddress(1, await wallet.getNewAddress()); // confirm it
  }

  // 9-10) Poll status until a terminal state (or a few iterations).
  let state: PaymentState | undefined;
  for (let i = 0; i < 10; i++) {
    state = await checkPayment(wallet, req);
    console.log(`status=${state.status} confirmations=${state.observation.confirmations}`);
    if (state.terminal) break;
    await new Promise((r) => setTimeout(r, 2000));
  }
  console.log("Final status:", state?.status);
}

main().catch((err) => {
  console.error("error:", err.message);
  process.exit(1);
});
